### DISCUSSION TITLE
Intellisense doesn't seem to recognize prototypes in copy books

### DISCUSSION THREAD

**Discussion by Gregory1701** [codefori/vscode-ibmi#1476] [IBM i] [answered]
In the attached video, I tried to demonstrate a simple copybook where I have a prototype.  Then I try to get intellisense to suggest that prototype for me.  This works in RDi (as demonstrated in the video), however, I cannot seem to get it to work in Code for i.  Am I missing an important setting or something?

On occasion (haven't been able to put my finger on it), code for i will suggest a procedure for me and I press enter - meaning to go to the next line, but instead it adds some procedure I don't want and it adds the prototype to my source.  This will cause a compiler error since the prototype is now in my source twice; once that was auto inserted and once in the copybook.  Is there a way to turn off the auto insert of the prototype?

I've also attached my code example.

Thanks in advance,
Gregory

https://github.com/halcyon-tech/vscode-ibmi/assets/132368707/22d7fbd9-1c16-489d-8fbe-493050111cbb
[MyPGM1.txt](https://github.com/halcyon-tech/vscode-ibmi/files/12268044/MyPGM1.txt)
[MyPGM1PR.txt](https://github.com/halcyon-tech/vscode-ibmi/files/12268045/MyPGM1PR.txt)

**Comment by worksofliam** [comment]
> however, I cannot seem to get it to work in Code for i.

So as of today, copybooks will only be picked up if it's the same library as the source you're editing. We have a PR to add library list support but it has not yet been merged. [You are welcome to try it](https://github.com/halcyon-tech/vscode-rpgle/pull/225#issuecomment-1664312523).

You can determine if the copybook has been resolved or not by hovering your mouse over the copy or include statement, and the hover text will tell you where it has been found or not.

> but instead it adds some procedure I don't want and it adds the prototype to my source.

VS Code will show auto-imports in the content assist, which automatically defines the prototype and adds the call to it if the prototype has not been defined before - this sounds like what you were experiencing. There is no way to turn them off, but you can escape the content assist with Escape to resume editing.

**Reply by Gregory1701** [reply, marked as answer]
Looking forward to the added library list support for copy books!

In my video example, the copybook is not only in my library (which is my current library), but it's also in the same source physical file.

It seems like, to me at least, that auto inserting a prototype that's already defined in the copybook is a waste.  Or at least, I'm not understanding why/when doing so would be a good thing.

**Reply by worksofliam** [reply]
Additional note: Looks like there is a bug where it is looking for `QRPGLEREF` when you aren't explicitly defining the source file, where it should be `QRPGLESRC`. This is already fixed as part of that PR I linked.

For example, currently:

* `/copy MYPGM1PR` resolved to `BASELIB/QRPGLEREF/MYPGM1PR` (the bug, fixed in the PR)
* `/copy qrpglesrc,mypgm1pr` resolved to `BASELIB/QRPGLESRC/MYPGM1PR`

**Reply by worksofliam** [reply]
@Gregory1701 

> but it's also in the same source physical file.

We must have replied at the same time, because I just discovered that bug ha. If you explicitly reference the source file for the time being, that should solve it.

**Reply by Gregory1701** [reply]
Well, I think I got the PR installed, but it's still not finding my procedure that's defined in the copybook.
![image](https://github.com/halcyon-tech/vscode-ibmi/assets/132368707/3749363d-c75d-408a-a271-b703eace98c7)
![image](https://github.com/halcyon-tech/vscode-ibmi/assets/132368707/ec2a73bc-bb29-4768-b6c8-953da2509ae7)

**Reply by worksofliam** [reply]
@Gregory1701 if you hover copy the copy directive, does it indicated found or not found?

I assume you have the RPGLE (via the PR vsix install) extension installed AND the language id, as indicated in the bottom right, is ‘rpgle’

**Reply by Gregory1701** [reply]
https://github.com/halcyon-tech/vscode-ibmi/assets/132368707/ab58f06a-928a-47a2-ba20-3744e602cd2e

**Reply by worksofliam** [reply]
To me it looks like something is missing from your setup - namely the RPGLE extension, which you should have if you installed the vsix from the PR.

**Reply by Gregory1701** [reply]
Maybe something has gotten messed up with my setup.  I'm noticing after installing the PR that now my outline view always reports 'No symbols found in document...'  How do you recommend I 'clean' or 'repair' my setup?

**Reply by Gregory1701** [reply]
@worksofliam it seems that every morning when I start my laptop from a cold start, launch VS Code and start to work I've once again lost the RPG extension from my WSL:Fedora instance.  I go up to my local extensions and click 'Install in WSL:Fedora' again and all is well.  Any thoughts on why I have to do this every day?

**Reply by manu-marcos** [reply]
Hello, has the fix been released yet? When I move the cursor over the copybook, it says found, but the content is not offered to me in autocomplete. I only have autocomplete when the copybook is in the same file as my program source.
